### PR TITLE
feat(iot): add Action to forward message attributes to Location

### LINKED ISSUE #24035: (iot): TopicRule action for Location
### Describe the feature

The CDK has no support to trigger Amazon Location via AWS IoT Rule.

### Use Case

Adding the action that sends location data to Amazon Location service directly through the AWS IoT topic rule.

### Proposed Solution

Implementing action class.

Probably, following PRs help coding.

* https://github.com/aws/aws-cdk/pull/17307
* https://github.com/aws/aws-cdk/pull/17466
* https://github.com/aws/aws-cdk/pull/17503

### Other Information

Following documentation will be help.

* https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-iot-topicrule-locationaction.html

It will be good first issue for learning implement feature in CDK!

Due to the bug https://github.com/aws/aws-cdk/issues/22732 it'll likely be required to override CloudFormation resources in the action class to handle the `Timestamp` field propely

### Acknowledgements

- [X] I may be able to implement this feature request
- [ ] This feature might incur a breaking change

### CDK version used

2.63.0

### Environment details (OS name and version, etc.)

MacOS, NodeJS 18

### PR DESCRIPTION
**Pull request by sthuber90**
Aims to add IoT action to integrate with Location service through L2 construct. 

The construct depends on `CfnTracker` from [aws_location](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_location-readme.html). I tried to import it with `import { CfnTracker } from '@aws-cdk/aws-location'` but without the following exclude in package.json `yarn build` failed. 

``` json
{
   // ...
   "pkglint": {
       "exclude": [
         "no-experimental-dependencies"
       ]
    }
}
```

According to the docs the location module is not experimental but there is a location-alpha module. So any info on how I need to import the location module to get rid of the include is greatly appreciated.

Closes #24035.

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-iot-actions/lib/location-action.ts
Comment (TheRealAmazonKendra) [inline packages/@aws-cdk/aws-iot-actions/lib/location-action.ts:55]: Can you help me understand how this is used? I had intended to suggest using `Duration` here, but if a timestamp is really the thing or something else, that's a bad suggestion on my end. Either way, can we make this more strongly typed than a string?

File: packages/@aws-cdk/aws-iot-actions/lib/location-action.ts
Comment (TheRealAmazonKendra) [inline packages/@aws-cdk/aws-iot-actions/lib/location-action.ts:76]: This gets converted to a string but it looks like they expect a double in the form of a string. Let's have users input a number and then we'll parse it for them into the string.

File: packages/@aws-cdk/aws-iot-actions/lib/location-action.ts
Comment (TheRealAmazonKendra) [inline packages/@aws-cdk/aws-iot-actions/lib/location-action.ts:84]: Same as above, please.

File: packages/@aws-cdk/aws-iot-actions/lib/location-action.ts
Comment (TheRealAmazonKendra) [inline packages/@aws-cdk/aws-iot-actions/lib/location-action.ts:128]: Would you be interested in contributing this fix before we get this merged?

File: packages/@aws-cdk/aws-iot-actions/test/location/integ.location-action.ts
Comment (TheRealAmazonKendra) [inline packages/@aws-cdk/aws-iot-actions/test/location/integ.location-action.ts:1]: Fantastic

File: packages/@aws-cdk/aws-iot-actions/lib/location-action.ts
Comment (TheRealAmazonKendra) [reply packages/@aws-cdk/aws-iot-actions/lib/location-action.ts:33]: Actually, can we use the `Duration` units here?

File: packages/@aws-cdk/aws-iot-actions/test/location/location-action.test.ts
Comment (TheRealAmazonKendra) [inline packages/@aws-cdk/aws-iot-actions/test/location/location-action.test.ts:1]: Not looking too closely at this just get given the changes I've requested.

**Review by TheRealAmazonKendra — CHANGES_REQUESTED** [review]
Thank you for your contribution! Overall this looks great, I just have some comments inline. 

For a second opinion, @yamatatsu, you've been leading the charge on IoT modules, could you take a look through this and provide feedback as well? I think you understand the service much better than I do.

File: packages/@aws-cdk/aws-iot-actions/lib/location-action.ts
Comment (sthuber90) [reply packages/@aws-cdk/aws-iot-actions/lib/location-action.ts:128]: I would be happy to but I haven't really understood how the types get generated and why this issue occurs. I'm happy to dig into it though

File: packages/@aws-cdk/aws-iot-actions/lib/location-action.ts
Comment (sthuber90) [reply packages/@aws-cdk/aws-iot-actions/lib/location-action.ts:33]: Thank you very much for your review. I will add some spaces.

I don't think that we can use `Duration` here. As I undersand it, `Duration` will provide a number but we actually require a string here. Also, the values allowed to be selected are very specific.

Bildschirm­foto 2023-02-10 um 09 20 00

File: packages/@aws-cdk/aws-iot-actions/lib/location-action.ts
Comment (sthuber90) [reply packages/@aws-cdk/aws-iot-actions/lib/location-action.ts:55]: We don't put a `Date` or `number` in here but an _expression_. This expression returns an epoch timestamp. See the example below. But in short: We can use the [IoT `timestamp()` function](https://docs.aws.amazon.com/iot/latest/developerguide/iot-sql-functions.html#iot-function-timestamp) here. Meaning we will have a vale `value: '${timestamp()}` which will be turned into the current epoch timestamp of when the message was processed.

File: packages/@aws-cdk/aws-iot-actions/lib/location-action.ts
Comment (sthuber90) [reply packages/@aws-cdk/aws-iot-actions/lib/location-action.ts:76]: From the documentation it says that we provide it as a string and it will be evaluated to a double value. More importantly, we need it to be a string as we can use the [IoT substitution template](https://docs.aws.amazon.com/iot/latest/developerguide/iot-substitution-templates.html). 

For example, if we receive a message such as the following on topic `dp/123456/data`:
```
{
   "latitude": 40.7484405,
   "longitude": -73.9856644,
}
```

Our SQL statement is just `SELECT ${topic(2} as device_id, * FROM dp/+/data`, then our IoT Location action will need to look like this:
```
{
      Location: {
        DeviceId: '${device_id}',
        Latitude: '${latitude}',
        Longitude: '${longitude}',
        RoleArn: 'theRoleArnToUseHere',
        Timestamp:  {
            Value: '${timestamp()}',
            Unit: LocationTimestampUnit.MILLISECONDS,
        },
        TrackerName: 'someTracker',
      },
}
```
in the console it would look as follows
Bildschirm­foto 2023-02-10 um 09 35 44

File: packages/@aws-cdk/aws-iot-actions/lib/location-action.ts
Comment (TheRealAmazonKendra) [reply packages/@aws-cdk/aws-iot-actions/lib/location-action.ts]: Hmm. For some reason I thought we had enum of the units in there. I was wrong, we don't.

File: packages/@aws-cdk/aws-iot-actions/lib/location-action.ts
Comment (TheRealAmazonKendra) [reply packages/@aws-cdk/aws-iot-actions/lib/location-action.ts:59]: Ah, I see. So there isn't really a good way to type this more strongly.

**Comment by sthuber90** [comment]
Cannot seem to get the PR fixed. The refactoring completely broke my local branch. I already checked out a new copy from the repository and ran `yarn install` from the root, ran `../../../scripts/buildup` from the _aws-iot-actions_ folder but that keeps failing with the following message during

```
/workspaces/aws-cdk/packages/@aws-cdk/aws-iot-actions: yarn build  (1 remaining)
yarn run v1.22.19
$ cdk-build
```

The error message
```
[2023-04-02T22:49:00.216] [ERROR] jsii/compiler - Compilation errors prevented the JSII assembly from being created
warning JSII6: A "peerDependency" on "constructs" at "^10.0.0" means you should take a "devDependency" on "constructs" at "10.0.0" (found "^10.0.0")
lib/location-action.ts:132:53 - error TS2724: '"/workspaces/aws-cdk/packages/@aws-cdk/aws-iot/lib/index"' has no exported member named 'CfnTopicRule'. Did you mean 'TopicRule'?

132     const topicRule = rule.node.defaultChild as iot.CfnTopicRule;
                                                        ~~~~~~~~~~~~
test/location/integ.location-action.ts:1:22 - error TS2307: Cannot find module '@aws-cdk/aws-iot' or its corresponding type declarations.

1 import * as iot from '@aws-cdk/aws-iot';
```

**Comment by sthuber90** [comment]
@yamatatsu would be great if we could merge this soon so that I don't have to update the PR after another refactor again

File: packages/@aws-cdk/aws-iot-actions/lib/location-action.ts
Comment (yamatatsu) [inline packages/@aws-cdk/aws-iot-actions/lib/location-action.ts:134]: This workaround will probably not work when using this action for second action of TopicRile because of `Actions.0`.

But I don’t know alternative…

**Review by yamatatsu — COMMENTED** [review]
@TheRealAmazonKendra 
Thanks for the mention!! And so sorry for the delay 🙇🏻.

@sthuber90
Thanks for the re-mention!! 🙏🏻 
I have made some comments including minor style comments, learning from comments I have received from maintainers in the past😉.

File: packages/@aws-cdk/aws-iot-actions/lib/location-action.ts
Comment (yamatatsu) [reply packages/@aws-cdk/aws-iot-actions/lib/location-action.ts:134]: It is maybe difficult to implement this action without fixing [this issue](https://github.com/aws/aws-cdk/issues/22732) I think...

File: packages/@aws-cdk/aws-iot-actions/lib/location-action.ts
Comment (sthuber90) [reply packages/@aws-cdk/aws-iot-actions/lib/location-action.ts:145]: Hope I moved this more into the direction that you envisioned. Hope to get rid of this override soon

File: packages/@aws-cdk/aws-iot-actions/lib/location-action.ts
Comment (sthuber90) [reply packages/@aws-cdk/aws-iot-actions/lib/location-action.ts:134]: Fully agree with you but also couldn't come up with a better workaround

Comment (sthuber90) [reply packages/@aws-cdk/aws-iot-actions/lib/location-action.ts:124]: Thank you very much for your review. I really appreciate all your feedback

File: packages/@aws-cdk/aws-iot-actions-alpha/lib/location-action.ts
Comment (corymhall) [inline packages/@aws-cdk/aws-iot-actions-alpha/lib/location-action.ts:103]: We don't want to expose any L1s in an L2. For this PR we will need to implement an L2 for CfnTracker as well.

**Review by corymhall — DISMISSED** [review]

File: packages/@aws-cdk/aws-iot-actions-alpha/lib/location-action.ts
Comment (sthuber90) [reply packages/@aws-cdk/aws-iot-actions-alpha/lib/location-action.ts:103]: I will try to add a L2 for the Location Tracker but will that be ok? The contribution guideline says that a new L2 must reference a RFC

Comment (corymhall) [reply packages/@aws-cdk/aws-iot-actions-alpha/lib/location-action.ts:103]: It looks like location tracker is a pretty simple resource and we already have an alpha module for location so we can skip the RFC.

**Review by corymhall — DISMISSED** [review]
putting back in request changes. Feel free to also change this to a draft if you don't plan on working on it soon.

**Pull request closed without merging**
